File: apps/trader-service/app/services/exchange_service.py
from __future__ import annotations

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.models import UserExchange
from app.utils.encryption import AesGcmEncryptor


class ExchangeService:
    def __init__(self, *, encryptor: AesGcmEncryptor | None):
        self._encryptor = encryptor

    def _require_encryptor(self) -> AesGcmEncryptor:
        if not self._encryptor:
            raise ValueError("ENCRYPTION_KEY is required for this operation")
        return self._encryptor

    async def list_for_user(self, session: AsyncSession, *, user_id: str) -> list[UserExchange]:
        result = await session.execute(
            select(UserExchange)
            .where(UserExchange.user_id == user_id)
            .order_by(UserExchange.created_at.desc())
        )
        return list(result.scalars().all())

    async def get(self, session: AsyncSession, *, user_id: str, exchange_id: str) -> UserExchange:
        exchange = await session.get(UserExchange, exchange_id)
        if not exchange or exchange.user_id != user_id:
            raise ValueError("Exchange not found")
        return exchange

    async def create(
        self,
        session: AsyncSession,
        *,
        user_id: str,
        exchange_type: str,
        name: str,
        api_key: str,
        api_secret: str,
        passphrase: str | None,
        is_testnet: bool,
        is_active: bool,
    ) -> UserExchange:
        encryptor = self._require_encryptor()
        exchange = UserExchange(
            user_id=user_id,
            exchange_type=exchange_type,
            name=name,
            api_key_encrypted=encryptor.encrypt(api_key),
            api_secret_encrypted=encryptor.encrypt(api_secret),
            passphrase_encrypted=encryptor.encrypt(passphrase) if passphrase else None,
            is_testnet=is_testnet,
            is_active=is_active,
        )
        session.add(exchange)
        await session.commit()
        await session.refresh(exchange)
        return exchange

    async def update(
        self,
        session: AsyncSession,
        *,
        user_id: str,
        exchange_id: str,
        name: str | None = None,
        api_key: str | None = None,
        api_secret: str | None = None,
        passphrase: str | None = None,
        is_testnet: bool | None = None,
        is_active: bool | None = None,
    ) -> UserExchange:
        exchange = await session.get(UserExchange, exchange_id)
        if not exchange or exchange.user_id != user_id:
            raise ValueError("Exchange not found")

        encryptor = self._require_encryptor()
        if name is not None:
            exchange.name = name
        if api_key is not None:
            exchange.api_key_encrypted = encryptor.encrypt(api_key)
        if api_secret is not None:
            exchange.api_secret_encrypted = encryptor.encrypt(api_secret)
        if passphrase is not None:
            exchange.passphrase_encrypted = encryptor.encrypt(passphrase) if passphrase else None
        if is_testnet is not None:
            exchange.is_testnet = is_testnet
        if is_active is not None:
            exchange.is_active = is_active

        await session.commit()
        await session.refresh(exchange)
        return exchange

    async def set_grpc_token(
        self,
        session: AsyncSession,
        *,
        user_id: str,
        exchange_id: str,
        token: str,
    ) -> UserExchange:
        exchange = await self.get(session, user_id=user_id, exchange_id=exchange_id)
        encryptor = self._require_encryptor()
        exchange.grpc_token_encrypted = encryptor.encrypt(token)
        await session.commit()
        await session.refresh(exchange)
        return exchange

    async def clear_grpc_token(
        self,
        session: AsyncSession,
        *,
        user_id: str,
        exchange_id: str,
    ) -> UserExchange:
        exchange = await self.get(session, user_id=user_id, exchange_id=exchange_id)
        exchange.grpc_token_encrypted = None
        await session.commit()
        await session.refresh(exchange)
        return exchange

    async def get_grpc_token(
        self,
        session: AsyncSession,
        *,
        user_id: str,
        exchange_id: str,
    ) -> str:
        exchange = await self.get(session, user_id=user_id, exchange_id=exchange_id)
        if not exchange.grpc_token_encrypted:
            raise ValueError("Exchange is not initialized (missing grpc token)")
        encryptor = self._require_encryptor()
        return encryptor.decrypt(exchange.grpc_token_encrypted)

    async def get_api_credentials(
        self,
        session: AsyncSession,
        *,
        user_id: str,
        exchange_id: str,
    ) -> tuple[str, str, str | None]:
        exchange = await self.get(session, user_id=user_id, exchange_id=exchange_id)
        encryptor = self._require_encryptor()
        api_key = encryptor.decrypt(exchange.api_key_encrypted)
        api_secret = encryptor.decrypt(exchange.api_secret_encrypted)
        passphrase = (
            encryptor.decrypt(exchange.passphrase_encrypted) if exchange.passphrase_encrypted else None
        )
        return api_key, api_secret, passphrase

    async def delete(self, session: AsyncSession, *, user_id: str, exchange_id: str) -> None:
        exchange = await self.get(session, user_id=user_id, exchange_id=exchange_id)
        await session.delete(exchange)
        await session.commit()

File: apps/trader-service-node/src/service/exchange.service.ts
import { Config, Provide, Scope, ScopeEnum, httpError } from '@midwayjs/core';
import { isNil } from 'lodash-es';
import { InjectEntityModel } from '@midwayjs/typeorm';
import type { Repository } from 'typeorm';
import { UserExchange } from '../entity/user-exchange.entity.js';
import { AesGcmEncryptor } from '../util/encryption.js';
import type { EncryptionConfig } from '../types/index.js';

type ExchangeCreateParams = {
  userid: number;
  exchangeType: string;
  name: string;
  apiKey: string;
  apiSecret: string;
  passphrase?: string;
  isTestnet?: boolean;
  isActive?: boolean;
};

type ExchangeUpdateParams = {
  exchangeType?: string;
  name?: string;
  apiKey?: string;
  apiSecret?: string;
  passphrase?: string;
  isTestnet?: boolean;
  isActive?: boolean;
};

@Provide()
@Scope(ScopeEnum.Request, { allowDowngrade: true })
export class ExchangeService {
  @InjectEntityModel(UserExchange)
  exchangeRepo?: Repository<UserExchange>;

  @Config('encryption')
  encryptionConfig!: EncryptionConfig;

  private requireRepo(): Repository<UserExchange> {
    if (!this.exchangeRepo) {
      throw new httpError.ServiceUnavailableError('Database not configured');
    }
    return this.exchangeRepo;
  }

  private requireEncryptor(): AesGcmEncryptor {
    const key = (this.encryptionConfig?.key ?? '').trim();
    if (!key) {
      throw new httpError.InternalServerErrorError('ENCRYPTION_KEY is required for this operation');
    }
    try {
      return AesGcmEncryptor.fromKey(key);
    } catch (err) {
      const message = err instanceof Error ? err.message : 'Invalid ENCRYPTION_KEY';
      throw new httpError.InternalServerErrorError(message);
    }
  }

  async listForUser(userid: number): Promise<UserExchange[]> {
    const repo = this.requireRepo();
    return repo.find({ where: { userid }, order: { createdAt: 'DESC' } });
  }

  async get(userid: number, exchangeId: number): Promise<UserExchange> {
    const repo = this.requireRepo();
    const exchange = await repo.findOne({ where: { id: exchangeId } });
    if (!exchange || exchange.userid !== userid) {
      throw new httpError.NotFoundError('Exchange not found');
    }
    return exchange;
  }

  async create(params: ExchangeCreateParams): Promise<UserExchange> {
    const repo = this.requireRepo();
    const encryptor = this.requireEncryptor();

    const exchange = repo.create({
      userid: params.userid,
      exchangeType: params.exchangeType,
      name: params.name,
      apiKeyEncrypted: encryptor.encrypt(params.apiKey),
      apiSecretEncrypted: encryptor.encrypt(params.apiSecret),
      passphraseEncrypted: params.passphrase ? encryptor.encrypt(params.passphrase) : null,
      isTestnet: params.isTestnet ?? false,
      isActive: params.isActive ?? true,
    });
    return await repo.save(exchange);
  }

  async update(userid: number, exchangeId: number, patch: ExchangeUpdateParams): Promise<UserExchange> {
    const repo = this.requireRepo();
    const encryptor = this.requireEncryptor();
    const exchange = await this.get(userid, exchangeId);

    if (!isNil(patch.name)) exchange.name = patch.name;
    if (!isNil(patch.apiKey)) {
      exchange.apiKeyEncrypted = encryptor.encrypt(patch.apiKey);
    }
    if (!isNil(patch.apiSecret)) {
      exchange.apiSecretEncrypted = encryptor.encrypt(patch.apiSecret);
    }
    if (!isNil(patch.exchangeType)) {
      exchange.exchangeType = patch.exchangeType;
    }
    if (patch.passphrase !== undefined) {
      exchange.passphraseEncrypted =
        patch.passphrase && patch.passphrase.trim() ? encryptor.encrypt(patch.passphrase) : null;
    }
    if (!isNil(patch.isTestnet)) exchange.isTestnet = patch.isTestnet;
    if (!isNil(patch.isActive)) exchange.isActive = patch.isActive;

    return await repo.save(exchange);
  }

  async delete(userid: number, exchangeId: number): Promise<void> {
    const repo = this.requireRepo();
    const exchange = await this.get(userid, exchangeId);
    await repo.remove(exchange);
  }

  async getApiCredentials(userid: number, exchangeId: number): Promise<{
    apiKey: string;
    apiSecret: string;
    passphrase: string | null;
  }> {
    const exchange = await this.get(userid, exchangeId);
    const encryptor = this.requireEncryptor();
    return {
      apiKey: encryptor.decrypt(exchange.apiKeyEncrypted),
      apiSecret: encryptor.decrypt(exchange.apiSecretEncrypted),
      passphrase: exchange.passphraseEncrypted ? encryptor.decrypt(exchange.passphraseEncrypted) : null,
    };
  }
}
